packages/modules/kostal_plenticore/device.py

```diff
@@ -0,0 +1,166 @@
+                        # Wird die Batterie entladen, werden die Wandlungsverluste anteilig an der DC-Leistung auf PV
+                        # und Batterie verteilt. Dazu muss der Divisor Total_DC_power != 0 sein.
+                        power_gross = dc_in - bat_state_net.power
+                        pv_state = InverterState(power=dc_in / power_gross * inverter_state.power,
+                                                 exported=inverter_state.exported)
+                        # Speicherladung muss durch Wandlungsverluste und internen Verbrauch korrigiert werden, sonst
+                        # wird ein falscher Hausverbrauch berechnet. Die Verluste fallen hier unter den Tisch.
+                        bat_state_gross = BatState(power=pv_state.power-inverter_state.power-home_consumption,
+                                                   imported=bat_state_net.imported,
+                                                   exported=bat_state_net.exported)
+                    else:
+                        # Wenn die Batterie geladen wird, dann ist PV-Leistung die Wechselrichter-AC-Leistung + die
+                        # Ladeleistung der Batterie. Die PV-Leistung ist die Summe aus verlustbehafteter
+                        # AC-Leistungsabgabe des WR und der DC-Ladeleistung. Die Wandlungsverluste werden also nur
+                        # in der PV-Leistung ersichtlich.
+                        pv_state = InverterState(power=inverter_state.power - bat_state_net.power,
+                                                 exported=inverter_state.exported)
+                        bat_state_gross = bat_state_net
+                    # https://github.com/snaptec/openWB/pull/2440#discussion_r996275286
+                    # power_gross = bat_state.power + dc_in
+                    # bat_state_gross = BatteryState(power=bat_state_net.power / power_gross * inverter_state.power)
+                    # pv_state = InverterState(power=inverter_state.power - bat_state_gross.power)
+                else:
+                    inverter_state.power = 0
+                    pv_state = inverter_state
+                    bat_state_gross = bat_state_net
+            else:
+                pv_state = inverter_state
+            if set_inverter_state:
+                component.update(pv_state)
+        elif isinstance(component, KostalPlenticoreCounter):
+            component.update(reader)
+    if bat_state_net:
+        battery.update(bat_state_gross)
+    if set_inverter_state is False:
+        return pv_state
+
+
+def create_device(device_config: KostalPlenticore):
+    def create_bat_component(component_config):
+        return KostalPlenticoreBat(device_config.id, component_config)
+
+    def create_counter_component(component_config):
+        return KostalPlenticoreCounter(device_config.id, component_config)
+
+    def create_inverter_component(component_config):
+        return KostalPlenticoreInverter(component_config)
```

Comment:
Der Parameter `component_config` in diesen drei Funktionen braucht noch eine Type-annotation. Das ist der Grund für den Fehler [hier im Forum](https://openwb.de/forum/viewtopic.php?p=75421#p75421)